From 8b89cb2e9b927de818cc950dcb237da5381665fe Mon Sep 17 00:00:00 2001
From: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
Date: Wed, 10 Aug 2022 23:47:57 -0700
Subject: [PATCH 11/12] ath12k: Make use of TID classification offload

Don't set HAL_TCL_DATA_CMD_INFO2_TID_OVERWRITE flag to TCL,
HW only take care of tid classification if this flag is not set.

Signed-off-by: Pradeep Kumar Chitrapu <quic_pradeepc@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_tx.c  | 17 -----------------
 drivers/net/wireless/ath/ath12k/hal_tx.c |  3 +--
 drivers/net/wireless/ath/ath12k/mac.c    |  1 +
 3 files changed, 2 insertions(+), 19 deletions(-)

--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -39,19 +39,6 @@ static void ath12k_dp_tx_encap_nwifi(str
 	hdr->frame_control &= ~__cpu_to_le16(IEEE80211_STYPE_QOS_DATA);
 }
 
-static u8 ath12k_dp_tx_get_tid(struct sk_buff *skb)
-{
-	struct ieee80211_hdr *hdr = (void *)skb->data;
-	struct ath12k_skb_cb *cb = ATH12K_SKB_CB(skb);
-
-	if (cb->flags & ATH12K_SKB_HW_80211_ENCAP)
-		return skb->priority & IEEE80211_QOS_CTL_TID_MASK;
-	else if (!ieee80211_is_data_qos(hdr->frame_control))
-		return HAL_DESC_REO_NON_QOS_TID;
-	else
-		return skb->priority & IEEE80211_QOS_CTL_TID_MASK;
-}
-
 enum hal_encrypt_type ath12k_dp_tx_get_encrypt_type(u32 cipher)
 {
 	switch (cipher) {
@@ -212,10 +199,6 @@ tcl_ring_sel:
 			     u32_encode_bits(1, HAL_TCL_DATA_CMD_INFO2_TCP6_CKSUM_EN);
 	}
 
-	ti.flags1 |= u32_encode_bits(1, HAL_TCL_DATA_CMD_INFO3_TID_OVERWRITE);
-
-	ti.tid = ath12k_dp_tx_get_tid(skb);
-
 	switch (ti.encap_type) {
 	case HAL_TCL_ENCAP_TYPE_NATIVE_WIFI:
 		ath12k_dp_tx_encap_nwifi(skb);
--- a/drivers/net/wireless/ath/ath12k/hal_tx.c
+++ b/drivers/net/wireless/ath/ath12k/hal_tx.c
@@ -60,8 +60,7 @@ void ath12k_hal_tx_cmd_desc_setup(struct
 		u32_encode_bits(ti->data_len, HAL_TCL_DATA_CMD_INFO2_DATA_LEN) |
 		u32_encode_bits(ti->pkt_offset, HAL_TCL_DATA_CMD_INFO2_PKT_OFFSET);
 
-	tcl_cmd->info3 = (ti->flags1 |
-		u32_encode_bits(ti->tid, HAL_TCL_DATA_CMD_INFO3_TID)) |
+	tcl_cmd->info3 = ti->flags1 |
 		u32_encode_bits(ti->lmac_id, HAL_TCL_DATA_CMD_INFO3_PMAC_ID) |
 		u32_encode_bits(ti->vdev_id, HAL_TCL_DATA_CMD_INFO3_VDEV_ID);
 
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -10929,6 +10929,7 @@ static int ath12k_mac_hw_register(struct
 	ieee80211_hw_set(hw, SUPPORTS_REORDERING_BUFFER);
 	ieee80211_hw_set(hw, SUPPORTS_AMSDU_IN_AMPDU);
 	ieee80211_hw_set(hw, USES_RSS);
+	ieee80211_hw_set(hw, SUPPORTS_TID_CLASS_OFFLOAD);
 
 	if (ath12k_frame_mode == ATH12K_HW_TXRX_ETHERNET) {
 		ieee80211_hw_set(hw, SUPPORTS_TX_ENCAP_OFFLOAD);
